src/Plugins/Loader/ManagedLoadContext.cs

```diff
@@ -88,11 +91,13 @@ internal class ManagedLoadContext : AssemblyLoadContext
                 // If default context is preferred, check first for types in the default context unless the dependency has been declared as private
                 try
                 {
-                    var defaultAssembly = Default.LoadFromAssemblyName(assemblyName);
+                    var defaultAssembly = _defaultLoadContext.LoadFromAssemblyName(assemblyName);
                     if (defaultAssembly != null)
                     {
-                        // return null so ALC will fallback to loading from Default ALC directly
-                        return null;
+                        // Older versions used to return null here such that returned assembly would be resolved from the default ALC.
+                        // However, with the addition of custom default ALCs, the Default ALC may not be the user's chosen ALC when
+                        // this context was built. As such, we simply return the Assembly from the user's chosen default load context.
+                        return defaultAssembly;
```

Comment:
This is the one change that I am concerned about. Which context will this instance of `Assembly` belong to? The reason for returning `null` was to ensure the default context is used, and I don't know what behavior we will get here. I'd like to see some automated tests added that validate this is actually working before we move forward with merging this.

Reply:
This is actually the exact line that makes this whole PR work.
In Core, assemblies and contexts have a 1:1 mapping to my understanding. This comes from recently added ALC documentation on MSDN and some manual debugging to confirm.

In other words, when loaded by an ALC, the `Assembly` instance belongs to that ALC. Just like plugin interfaces, having the object cross ALCs doesn't mean that the object will be loaded into the foreign ALC.

You should be able to verify this in a debugger by creating an ALC "B" and using `LoadFromAssemblyName`, manually with that ALC. (Note: Create the `AssemblyName` manually without loading the assembly). Then, once that Assembly is loaded you can verify the assemblies loaded into a context using the `Assemblies` property of an ALC. The assembly will not be in the current context, only in ALC "B"'s Assemblies list.

Also random note: I noticed this after I made the PR but this MSDN page exists: (https://docs.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support)[Creating App with Plugin Support] , it does pretty much the same thing as this PR.

Reply:
This is actually the exact line that makes this whole PR work.
In Core, assemblies and contexts have a 1:1 mapping to my understanding. This comes from recently added ALC documentation on MSDN and some manual debugging to confirm.

In other words, when loaded by an ALC, the Assembly instance belongs to that ALC. Just like plugin interfaces, having the object cross ALCs doesn't mean that the object will be loaded into the foreign ALC.

You should be able to verify this in a debugger by creating an ALC "B" and using LoadFromAssemblyName, manually with that ALC. (Note: Create the AssemblyName manually without loading the assembly). Then, once that Assembly is loaded you can verify the assemblies loaded into a context using the Assemblies property of an ALC. The assembly will not be in the current context, only in ALC "B"'s Assemblies list.

Also random note: I noticed this after I made the PR but this MSDN page exists: [Creating App with Plugin Support](https://docs.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support) , it does pretty much the same thing as this PR.